### PR TITLE
randomized response bitvec FFI

### PR DESCRIPTION
**Pull request by Shoeboxam**
* adds support for passing bit vectors over FFI
* adds FFI for make_randomized_response_bitvec and its post-processor

FFI is functional for both Python <-> Rust communication and R <-> Rust communication.

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/1680?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1680** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1680?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1279** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1279?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: R/opendp/tests/testthat/test-measurements.R [hunk 1]
```diff
@@ -239,3 +239,15 @@ test_that("test_gaussian", {
...
+
+
+test_that("test_randomized_response_bitvec", {
+  xx <- raw(3)
+  xx[1] <- charToRaw("A")
+  xx[2] <- as.raw(1L)
+  input_space <- c(bitvector_domain(max_weight = 4L), discrete_distance())
+  m_rr <- input_space |> then_randomized_response_bitvec(f = 0.95)
+
+  expect_equal(m_rr(d_in = 1L), 0.8006676684558611)
```
Comment (mccalluc) [inline R/opendp/tests/testthat/test-measurements.R:251]: Add a comment with the math that produces this value, or get the value by evaluating a less opaque expression?

File: python/src/opendp/_convert.py [hunk 2]
```diff
@@ -320,6 +326,26 @@ def _slice_to_string(raw: FfiSlicePtr) -> str:
...
+    if not isinstance(val, (bytes, bytearray)):
+        raise TypeError("Expected type is BitVector but input data is not bytes or bytearray.")
+
+    array = (ctypes.c_uint8 * len(val)).from_buffer_copy(val) # type: ignore[operator]
+    return _wrap_in_slice(array, len(val) * 8)
+
+
+def _slice_to_bitvector(raw: FfiSlicePtr) -> bytes:
+    # raw.contents.len is the number of valid bits.
+    # ceiling division by 8 gives the number of bytes in the buffer
```
Comment (mccalluc) [inline python/src/opendp/_convert.py:343]:
```suggestion
    # Division by -8 is ceiling rather than floor: the number of bytes in the buffer
```

File: python/test/test_convert.py [hunk 3]
```diff
@@ -160,3 +160,14 @@ def test_polars_expr():
...
+
+def test_bitvec():
+    np = pytest.importorskip('numpy') 
+
+    for i in range(1, 20):
+        data = np.packbits([1] * i)
+        obj = py_to_c(data.tobytes(), AnyObjectPtr, "BitVector")
+        val_out = np.frombuffer(c_to_py(obj), dtype=np.uint8)
+        bits = np.unpackbits(val_out)
+        assert (bits.tolist() + [0]).index(0) == i
```
Comment (mccalluc) [inline python/test/test_convert.py:173]:
```suggestion
        assert len(bits.tolist()) == i
        assert all(x == 1 for x in bits.tolist())
```
Perhaps easier to read?

File: python/test/test_meas.py [hunk 4]
```diff
@@ -221,3 +221,21 @@ def test_alp_histogram():
...
+    np = pytest.importorskip('numpy')
+
+    m_rr = dp.m.make_randomized_response_bitvec(
+        dp.bitvector_domain(max_weight=4), dp.discrete_distance(), f=0.95
+    )
+
+    data = np.packbits(
+        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0]
+    )
+    assert np.array_equal(data, np.array([0, 8, 12], dtype=np.uint8))
```
Comment (mccalluc) [inline python/test/test_meas.py:235]: Checking my comprehension, you could add more zero-bits to the end without changing `data`?

File: python/test/test_meas.py [hunk 5]
```diff
@@ -221,3 +221,21 @@ def test_alp_histogram():
...
+
+    # roundtrip: bytes -> mech -> numpy
+    release = np.frombuffer(m_rr(data.tobytes()), dtype=np.uint8)
+
+    print(np.unpackbits(data), np.unpackbits(release))
+    assert m_rr.map(1) == 0.8006676684558611
```
Comment (mccalluc) [inline python/test/test_meas.py:241]: again, either having this as an expression, or having math that can be checked by hand would be good. Someone coming back to this in the future will only know that this is the value that passes the test, but it will take some time to understand where it comes from.

**Review by mccalluc — COMMENTED** [review]
Where there are particular floats in the tests, could you either define them with some expression, or provide the math in a comment? That would make the tests more informative, and easier to update if needed.

File: python/test/test_convert.py (hunk 3 above)
Comment (Shoeboxam) [reply python/test/test_convert.py:173]: The bit vector might be a mixture of zeros and ones-- this checks that the first zero bit is i.

File: python/test/test_meas.py (hunk 4 above)
Comment (Shoeboxam) [reply python/test/test_meas.py:235]: Yes, the input is always padded out with zeros.

File: R/opendp/src/convert.c [hunk 6]
```diff
@@ -93,11 +93,15 @@ void *sexp_to_voidptr(SEXP input, SEXP rust_type)
...
+    if (str_equal(sexp_to_charptr(get_origin(rust_type)), "Option"))
+    {
+        if (input == R_NilValue)
+        {
             UNPROTECT(2);
             return NULL;
-        } else {
+        }
+        else
+        {
```
Comment (mccalluc) [inline R/opendp/src/convert.c:104]: Formatting changes don't seem to be related to PR... but not causing a problem either.

File: python/src/opendp/_convert.py [hunk 7]
```diff
@@ -320,6 +326,26 @@ def _slice_to_string(raw: FfiSlicePtr) -> str:
     return value.decode()
 
 
+def _bitvector_to_slice(val: Sequence[Any]) -> FfiSlicePtr:
+    np = import_optional_dependency('numpy', raise_error=False)
+    if np is not None and isinstance(val, np.ndarray):
+        val = val.tobytes()
+    
+    if not isinstance(val, (bytes, bytearray)):
+        raise TypeError("Expected type is BitVector but input data is not bytes or bytearray.")
```
Comment (mccalluc) [inline python/src/opendp/_convert.py:335]: Since this function should work with or w/o numpy, you could add an test of the behavior w/o numpy... but the numpy route exercises all the lines, so not critical.

**Review by mccalluc — APPROVED** [review]
A couple comments, but fine to merge as-is. Thanks!
